applications/StructuralMechanicsApplication/custom_utilities/multipoint_constraint_data.hpp

```diff
@@ -0,0 +1,246 @@
+    {
+        return this->slaveDofMasterDataMap[slaveDOF].masterWeights;
+    }
+
+    /**
+		Get the Total number of MasterDOFs in the current Node
+		@return Total number of MasterDOFs in the current Node
+		 */
+    const int GetTotalNumbeOfMasterDOFs()
+    {
+        this->totalNumberOfMasterDOFs = 0;
+        for (const auto it : this->slaveDofMasterDataMap)
+        {
+            this->totalNumberOfMasterDOFs += (it.second).masterDOFs.size();
+        }
+        return this->totalNumberOfMasterDOFs;
+    }
+
+    /**
+		Get the SlaveDOFs vector for this Node
+		@return SlaveDOFs vector for this Node
+		*/
+    const std::vector<DofPointerType> GetSlaveDOFs()
+    {
+        return this->slaveDofVector;
+    }
+
+    ///@
+
+    ///@name Static Operations
+    ///@{
+    /**
+		 * Returns the string containing a detailed description of this object.
+		 * @return the string with informations
+		 */
+    virtual void GetInfo() const
+    {
+        std::cout << std::endl;
+        std::cout << "===============================================================" << std::endl;
+        std::cout << "Total number of slaves :: " << (slaveDofVector).size() << std::endl;
+        for (const auto it : this->slaveDofMasterDataMap)
+        {
+            std::cout << "SlaveDOF :: " << (it.first)->EquationId() << std::endl;
+            for (unsigned int i = 0; i < (it.second).masterDOFs.size(); ++i)
+            {
+                std::cout << "\t MasterDOFs :: " << ((it.second).masterDOFs[i])->EquationId() << std::endl;
+                std::cout << "\t MasterWeights ::" << (it.second).masterWeights[i] << std::endl;
+            }
+        }
+        std::cout << "===============================================================" << std::endl;
+        std::cout << std::endl;
+    }
+
+    ///@}
+    virtual void PrintInfo(std::ostream &rOStream) const
+    {
+        rOStream << " MpcData object " << std::endl;
+    }
+
+  private:
+    ///@name Member Variables
+    ///@{
+    std::vector<DofPointerType> slaveDofVector;
+    unsigned int totalNumberOfMasterDOFs;
+
+    struct masterData
+    {
+        std::vector<DofPointerType> masterDOFs;
+        std::vector<double> masterWeights;
+    };
+
+    std::map<DofPointerType, masterData> slaveDofMasterDataMap;
+    ///@}
+
+    ///@name Serialization
+    ///@{
+    friend class Serializer;
+
+    virtual void save(Serializer &rSerializer) const
+    {
+        KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, Flags);
+    }
+
+    virtual void load(Serializer &rSerializer)
+    {
+        KRATOS_SERIALIZE_LOAD_BASE_CLASS(rSerializer, Flags);
+    }
+
+    ///@}
+};
+
+///@name Input/Output funcitons
+///@{
+
+inline std::istream &operator>>(std::istream &rIStream, MpcData &rThis);
+
+inline std::ostream &operator<<(std::ostream &rOStream, const MpcData &rThis)
+{
+    return rOStream;
+}
+
+///@}
+
+} // namespace Kratos
+
+#endif // CONSTRAINT_SLAVE_H_INCLUDED

```

Comment:
Further thinking on this, why instead of having one MPC per node, we create a global map

MPCs[slave_dof] vs master_dofs & weights?

this could be one single variable _of the model part_  instead of being assigned to the nodes, and if we make it as a unordered_map it would allow fast access